pkg/reconcilers/external/kubeconfig.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	key := util.DefaultString(hcp.Spec.BootstrapSecretRef.Key, util.KubeconfigSecretKeyInCluster)
+
+	kconfigBytes := bootstrapSecret.Data[key]
+	if kconfigBytes == nil {
+		return nil, fmt.Errorf("kubeconfig not found in bootstrap secret for key %s", key)
+	}
+
+	return clientcmd.Load(kconfigBytes)
+}
+
+func reconcileAdoptedClusterRoleBinding(clientset *kubernetes.Clientset, saName, saNamespace string) error {
+	name := saName + "-clusterrolebinding"
+
+	// Check if the ClusterRoleBinding already exists
+	_, err := clientset.RbacV1().ClusterRoleBindings().Get(context.TODO(), name, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Define new ClusterRoleBinding
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      saName,
+				Namespace: saNamespace,
+			},
+		},
+	}
+
+	// Create the ClusterRoleBinding
+	_, err = clientset.RbacV1().ClusterRoleBindings().Create(context.TODO(), clusterRoleBinding, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func reconcileAdoptedServiceAccount(saClient v1.ServiceAccountInterface, saName string) error {
+
+	// Check if the ServiceAccount already exists
+	_, err := saClient.Get(context.TODO(), saName, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Define a new ServiceAccount
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: saName,
+		},
+	}
+
+	// Create the ServiceAccount
+	_, err = saClient.Create(context.TODO(), sa, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func requestTokenWithExpiration(saClient v1.ServiceAccountInterface, saName string, expirationSeconds *int64) (string, error) {
+	if expirationSeconds == nil {
+		expirationSeconds = &defaultAdoptedTokenExpirationSeconds
+	}
+
+	tokenRequest := &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
+			Audiences:         []string{},
+			ExpirationSeconds: expirationSeconds,
+		},
+	}
+
+	tokenResponse, err := saClient.CreateToken(context.TODO(), saName, tokenRequest, metav1.CreateOptions{})
+	if err != nil {
+		return "", err
+	}
+	return tokenResponse.Status.Token, nil
+}
+
+func createNewKubeConfig(bootstrapConfig *api.Config, token string, hcp *tenancyv1alpha1.ControlPlane) ([]byte, error) {
+	context := bootstrapConfig.Contexts[bootstrapConfig.CurrentContext]
+
+	cluster := bootstrapConfig.Clusters[context.Cluster]
+	if cluster == nil {
+		return nil, fmt.Errorf("invalid cluster name %s for adoped cluster", context.Cluster)
+	}
+
+	kubeConfig := api.NewConfig()
+
+	kubeConfig.Clusters[context.Cluster] = &api.Cluster{
```

Comment:
Aren't there additional fields of `cluster` that should be copied? Is there a reason to not copy `cluster` as a whole rather than a few enumerated fields?